internal/pkg/manifest/union.go

```diff
@@ -0,0 +1,156 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"gopkg.in/yaml.v3"
+)
+
+// Union is a type used for yaml keys that may be of type Simple or Advanced.
+// Union will only ever hold one of the underlying types, never both.
+//
+// Use NewUnionSimple() and NewUnionAdvanced() to create a Union with an underlying
+// type already set. See Unmarshal() for details on how yaml is decoded
+// into Union.
+//
+// Union is exported to enable type embedding.
+type Union[Simple, Advanced any] struct {
+	// isSimple is true if the underlying type of Union is Simple.
+	isSimple bool
+
+	// Simple holds the value of Union if IsSimple() is true.
+	// If IsSimple() is false, this is the zero value of type Simple.
+	//
+	// Simple is exported to support mergo. It should not be set
+	// directly. Use NewUnionSimple() to create a Union with Simple set.
+	Simple Simple
+
+	// isAdvanced is true if the underlying type of Union is Advanced.
+	isAdvanced bool
```

Comment:
isn't `isAdvanced = !isSimple`? do we need this field

Reply:
clever clever 😏

Reply:
actually jk. we need it😛. there are three states:
1. Underlying type is `Simple`
2. Underlying type is `Advanced`
3. Underlying type is neither

we could represent with a single field (int enum or *bool) instead, though i think i prefer this option 🤔